triggers/daml/Daml/Trigger/LowLevel.daml

```diff
@@ -244,3 +259,94 @@ data Commands = Commands
   { commandId : CommandId
   , commands : [Command]
   }
+
+newtype StateT s m a = StateT { runStateT : s -> m (a, s) }
+  deriving Functor
+
+liftStateT : Functor m => m a -> StateT s m a
+liftStateT ma = StateT $ \s -> (,s) <$> ma
+
+instance Action m => Applicative (StateT s m) where
+  pure a = StateT (\s -> pure (a, s))
+  f <*> x = f >>= (<$> x)
+  
+instance Action m => Action (StateT s m) where
+  StateT x >>= f = StateT $ \s -> do
+    (x', s') <- x s
+    runStateT (f x') s'
+
+class ActionState s m | m -> s where
+  get : m s
+  put : s -> m ()
+
+modify : (Action m, ActionState s m) => (s -> s) -> m ()
+modify f = put . f =<< get
+
+instance Applicative m => ActionState s (StateT s m) where
+  get = StateT $ \s -> pure (s, s)
+  put s = StateT $ const $ pure ((), s)
+
+execStateT : Functor m => StateT s m a -> s -> m s
+execStateT (StateT fa) = fmap snd . fa
+
+zoom : Functor m => (t -> s) -> (t -> s -> t) -> StateT s m a -> StateT t m a
+zoom r w (StateT smas) = StateT $ \t ->
+  smas (r t) <&> \(a, s) -> (a, w t s)
+
+-- Must be kept in sync with Runner#handleStepFreeResult
+data TriggerF a =
+  GetTime (Time -> a)
+  | Submit (Commands, () -> a)
+  deriving Functor
+
+newtype TriggerSetup a = TriggerSetup { runTriggerSetup : Free TriggerF a }
+  deriving (Functor, Applicative, Action)
+
+newtype TriggerRule s a = TriggerRule { runTriggerRule : StateT s (Free TriggerF) a }
+  deriving (Functor, Applicative, Action)
+
+-- | Run a rule without running it.  May lose information from the rule;
+-- meant for testing purposes only.
+simulateRule : TriggerRule s a -> Time -> s -> (s, [Commands], a)
+simulateRule rule time s = (s', reverse cmds, a)
+  where ((a, s'), cmds) = State.runState (foldFree sim (runStateT (runTriggerRule rule) s)) []
+        sim : TriggerF x -> State.State [Commands] x
+        sim (GetTime f) = pure (f time)
+        sim (Submit (cmds, f)) = State.modify (cmds ::) $> f ()
+
+deriving instance ActionState s (TriggerRule s)
+
+class HasTime m => ActionTrigger m where
+  liftTF : TriggerF a -> m a
+
+instance ActionTrigger TriggerSetup where
+  liftTF = TriggerSetup . lift
+
+instance ActionTrigger (TriggerRule s) where
+  liftTF = TriggerRule . liftStateT . lift
+
+instance HasTime TriggerSetup where
+  getTime = liftTF (GetTime identity)
+
+instance HasTime (TriggerRule s) where
+  getTime = liftTF (GetTime identity)
+
+submitCommands : ActionTrigger m => Commands -> m ()
+submitCommands cmds = liftTF (Submit (cmds, identity))
+
+type TriggerInitialState s = Party -> Time -> ActiveContracts -> (s, [Commands])
+type TriggerUpdate s = Time -> Message -> s -> (s, [Commands])
+
+forwardPortInitialState : TriggerInitialState s -> Party -> ActiveContracts -> TriggerSetup s
+forwardPortInitialState f p ac = do
+  time <- getTime
+  let (s, cmds) = f p time ac
+  mapA_ submitCommands cmds
+  pure s
+
+forwardPort : TriggerUpdate s -> Message -> TriggerRule s ()
```

Comment:
It looks like we only use this in 5 tests. 4 of those don’t use `getTime` so the time aspect becomes redundant. Given that, I would prefer to just port those tests instead of keeping this around forever. But doesn’t have to be in this PR.

Reply:
1187207430d0c361188d55c39cb19c354fe4e084